README.md
# format-stripe-metadata [![npm](https://img.shields.io/npm/v/format-stripe-metadata.svg)](https://www.npmjs.com/package/format-stripe-metadata) [![npm](https://img.shields.io/npm/l/format-stripe-metadata.svg)](https://github.com/sublimefund/format-stripe-metadata/blob/master/LICENSE) [![Build Status](https://travis-ci.org/sublimefund/format-stripe-metadata.svg?branch=master)](https://travis-ci.org/sublimefund/format-stripe-metadata) [![codecov](https://codecov.io/gh/sublimefund/format-stripe-metadata/branch/master/graph/badge.svg)](https://codecov.io/gh/sublimefund/format-stripe-metadata)
[Stripe](https://stripe.com/) allows attaching
[metadata](https://stripe.com/docs/api#metadata) to certain objects. However,
there are documented limits on the metadata, the data must be in a key/value
format, and values must be strings (nested objects aren't allowed). If these
requirements aren't met, then calling the Stripe API will fail. This package
formats the input data to ensure compliance.

From the docs:
> You can specify up to 20 keys, with key names up to 40 characters long and values up to 500 characters long.

These values are used internally, but you can override them if necessary.

## Installation
```sh
$ yarn add format-stripe-metadata
# or
$ npm install --save format-stripe-metadata
```

## Usage
The returned object will contain a property called `_fullData` with a value of
either `'true'` or `'false'`. If the value is `'false'`, then information was
lost in the formatting process. Depending on how you are using the metadata
field, you may want to treat this situation as a bug, log a warning, etc.

Values are converted to strings as necesssary using [`JSON.stringify`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify).

```js
import {formatStripeMetadata} from 'format-stripe-metadata';
// or
const formatStripeMetadata = require('format-stripe-metadata').formatStripeMetadata;

let formatted = formatStripeMetadata(null);
// {_fullData: 'true'}

// Strings, numbers, and arrays are set as the value for the 'data' key
formatted = formatStripeMetadata('Mandarin orange');
// {_fullData: 'true', data: 'Mandarin orange'}
formatted = formatStripeMetadata(679);
// {_fullData: 'true', data: '679'}
formatted = formatStripeMetadata(['foo', 'bar', 'baz']);
// {_fullData: 'true', data: '["foo","bar","baz"]'}

// Objects are copied as much as possible
formatted = formatStripeMetadata({thomas: 'edison', nikola: 'tesla'});
// {_fullData: 'true', thomas: 'edison', nikola: 'tesla'}
formatted = formatStripeMetadata({thisisakeythatisabsurdlylong: 'Too Long'});
// {_fullData: 'false', thisisakeythatisabsu: 'Too Long'}

// The default limits can be overridden
formated = formatStripeMetadata(metadata, {
    maxNumKeys: 15,
    maxKeyLength: 18,
    maxValueLength: 400
});
```

## License
[MIT](https://github.com/sublimefund/format-stripe-metadata/blob/master/LICENSE)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).

## [Unreleased]

## 0.1.0 - 2018-03-26
### Added
- Initial implementation

[Unreleased]: https://github.com/sublimefund/format-stripe-metadata/compare/v0.1.0...HEAD


test.ts
import {
    formatStripeMetadata as format,
    MAX_KEY_LENGTH,
    MAX_NUM_KEYS,
    MAX_VALUE_LENGTH
} from './index';

test('empty values', () => {
    const emptyResult = {_fullData: 'true'};
    expect(format(null)).toEqual(emptyResult);
    expect(format(undefined)).toEqual(emptyResult);
    expect(format()).toEqual(emptyResult);
});

test('string', () => {
    expect(format('Why Walk a Dog?')).toEqual({
        _fullData: 'true',
        data: 'Why Walk a Dog?'
    });
});

test('number', () => {
    expect(format(3456)).toEqual({
        _fullData: 'true',
        data: '3456'
    });
});

test('array', () => {
    expect(format([])).toEqual({_fullData: 'true', data: '[]'});
});

test('valid object', () => {
    const metadata = {
        foo: 3,
        bar: 'asdf',
        baz: false
    };

    expect(format(metadata)).toEqual({
        _fullData: 'true',
        foo: '3',
        bar: 'asdf',
        baz: 'false'
    });
});

test('key is too long', () => {
    const key = 'a'.repeat(MAX_KEY_LENGTH + 1);
    const metadata = {
        [key]: 'Nowhere Fast'
    };

    expect(format(metadata)).toEqual({
        _fullData: 'false',
        [key.substr(0, MAX_KEY_LENGTH)]: 'Nowhere Fast'
    });
});

test('value is too long', () => {
    const value = 'a'.repeat(MAX_VALUE_LENGTH + 1);
    const metadata = {
        foo: value
    };

    expect(format(metadata)).toEqual({
        _fullData: 'false',
        foo: value.substr(0, MAX_VALUE_LENGTH)
    });
});

test('too many keys', () => {
    const metadata = {};
    for (const letter of 'abcdefghijklmnopqrstuvwxyz') {
        metadata[letter] = 'Once In My Life';
    }

    const formatted = format(metadata);
    expect(formatted._fullData).toBe('false');
    expect(Object.keys(formatted)).toHaveLength(MAX_NUM_KEYS);
});

test('custom max num keys', () => {
    const options = {maxNumKeys: 3};
    const metadata = {a: null, b: null, c: null, d: null, e: null};
    const formatted = format(metadata, options);
    expect(formatted._fullData).toBe('false');
    expect(Object.keys(formatted)).toHaveLength(options.maxNumKeys);
});

test('custom max key length', () => {
    const options = {maxKeyLength: 12};
    const metadata = {abcdefghijklmno: 'Severed'};
    const formatted = format(metadata, options);
    expect(formatted._fullData).toBe('false');
    expect(formatted).toHaveProperty('abcdefghijkl', 'Severed');
});

test('custom max value length', () => {
    const options = {maxValueLength: 8};
    const metadata = {foo: 'The meaning of life is'};
    const formatted = format(metadata, options);
    expect(formatted._fullData).toBe('false');
    expect(formatted).toHaveProperty('foo', 'The mean');
});


index.ts
import {IMetadata} from 'stripe';

// Default values from https://stripe.com/docs/api#metadata
export const MAX_KEY_LENGTH = 40;
export const MAX_NUM_KEYS = 20;
export const MAX_VALUE_LENGTH = 500;

export function formatStripeMetadata(
    data: any,
    options?: {
        maxKeyLength?: number;
        maxNumKeys?: number;
        maxValueLength?: number;
    }
): IMetadata {
    if (data === null || typeof data === 'undefined') {
        return {
            _fullData: 'true'
        };
    }

    const maxKeyLength =
        options && options.maxKeyLength ? options.maxKeyLength : MAX_KEY_LENGTH;
    const maxNumKeys =
        options && options.maxNumKeys ? options.maxNumKeys : MAX_NUM_KEYS;
    const maxValueLength =
        options && options.maxValueLength
            ? options.maxValueLength
            : MAX_VALUE_LENGTH;

    if (typeof data === 'string' || typeof data === 'number') {
        const dataString = data.toString();
        return {
            _fullData: (dataString.length <= maxValueLength).toString(),
            data: dataString.substr(0, maxValueLength)
        };
    }

    if (Array.isArray(data)) {
        const dataString = JSON.stringify(data);

        return {
            _fullData: (dataString.length <= maxValueLength).toString(),
            data: dataString.substr(0, maxValueLength)
        };
    }

    const keys = Object.keys(data);
    let fullData = keys.length <= maxNumKeys - 1;
    const metadata: IMetadata = {};

    for (const key of keys.slice(0, maxNumKeys - 1)) {
        let newKey = key;
        if (key.length > maxKeyLength) {
            newKey = key.substr(0, maxKeyLength);
            fullData = false;
        }

        let value =
            typeof data[key] === 'string'
                ? data[key]
                : JSON.stringify(data[key]);
        if (value.length > maxValueLength) {
            value = value.substr(0, maxValueLength);
            fullData = false;
        }

        metadata[newKey] = value;
    }

    metadata._fullData = fullData.toString();
    return metadata;
}
